sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -395,7 +395,14 @@ public void onResponse(Row response) {
 
                 @Override
                 public void onError(Throwable t) {
-                  future.setException(t);
+                  if (byteLimitReached) {
+                    // When the byte limit is reached we cancel the stream in onResponse.
+                    // In this case we don't want to fail the request with cancellation
+                    // exception. Instead, we construct the next request.
+                    onComplete();
```

Comment:
It seems off to reroute the exception to oncomplete, instead of handling it in onerror.

Will there be a scenario where we have reached the byte limit, but onComplete calls instead of onError?
Alternately, will there be a scenario where we have reached the byte limit, but there is another exception?

Reply:
> Will there be a scenario where we have reached the byte limit, but onComplete calls instead of onError?
I don't think so, I think the underlying client will bubble it up as a cancellation error.

> Alternately, will there be a scenario where we have reached the byte limit, but there is another exception?
Good point, I added an additional check to make sure the root cause is cancellation and only reroute in that case.

Reply:
Actually, once the byte limit is reached, we don't care if there are other errors, since we already buffered all the responses we care, so it's ok to jump to onComplete().

Reply:
wouldn't that be data loss?

Reply:
no, because in onComplete we construct the next request.

Reply:
What happens if we hit the byte limit, and in doing so we get an exception? wouldnt that mean that one of the records in the prior request is bad?

Reply:
I don't think that could happen. If there's a bad response, the exception will be thrown before onResponse() is called.

Reply:
got it. That makes sense to me

Reply:
In general, here is an expected scenario for server streaming rpcs:
- client consumes all of the messages, but before it receives the grpc trailers with an ok status, the deadline timer expires.
In this case, there is no reason to throw an error, as the customer got all of their data.

In this particular case, there is no data loss as the next segment read will because the next segment read will start from the last complete row (ie truncateRequest() will start from the last complete row. So calling onComplete when byteLimitReached is correct.


However there is another bug that I noticed: if a segment returns an empty result, then you will end up with index out of bounds exception

Reply:
chatted offline with Igor, we don't think that will be the case.